.gitlab-ci.yml

```diff
@@ -1,37 +1,591 @@
-image: docker:git
+
+variables:
+  MENDER_REV: "master"
+  META_MENDER_REV: "master"
+  POKY_REV: "thud"
+  MENDER_QA_REV: "master"
+  MENDER_ARTIFACT_REV: "master"
+  MENDER_CONDUCTOR_REV: "master"
+  MENDER_CONDUCTOR_ENTERPRISE_REV: "master"
+  DEPLOYMENTS_REV: "master"
+  DEVICEADM_REV: "master"
+  DEVICEAUTH_REV: "master"
+  GUI_REV: "master"
+  INVENTORY_REV: "master"
+  USERADM_REV: "master"
+  MENDER_API_GATEWAY_DOCKER_REV: "master"
+  MENDER_CLI_REV: "master"
+  INTEGRATION_REV: "master"
+  TENANTADM_REV: "master"
+  MENDER_STRESS_TEST_CLIENT_REV: "master"
+  META_OPENEMBEDDED_REV: "thud"
+  META_RASPBERRYPI_REV: "thud"
+  MENDER_IMAGE_TESTS_REV: "master"
+  BASE_BRANCH: "master"
+  BUILD_QEMUX86_64_UEFI_GRUB: "true"
+  TEST_QEMUX86_64_UEFI_GRUB: "true"
+  BUILD_QEMUX86_64_BIOS_GRUB: "true"
+  TEST_QEMUX86_64_BIOS_GRUB: "true"
+  BUILD_QEMUX86_64_BIOS_GRUB_GPT: "true"
+  TEST_QEMUX86_64_BIOS_GRUB_GPT: "true"
+  BUILD_VEXPRESS_QEMU_UBOOT_UEFI_GRUB: "true"
+  TEST_VEXPRESS_QEMU_UBOOT_UEFI_GRUB: "true"
+  BUILD_VEXPRESS_QEMU: "true"
+  TEST_VEXPRESS_QEMU: "true"
+  BUILD_VEXPRESS_QEMU_FLASH: "true"
+  TEST_VEXPRESS_QEMU_FLASH: "true"
+  BUILD_BEAGLEBONEBLACK: "true"
+  TEST_BEAGLEBONEBLACK: "false"
+  BUILD_RASPBERRYPI3: "false"
+  TEST_RASPBERRYPI3: "false"
+  RUN_INTEGRATION_TESTS: "true"
+  SPECIFIC_INTEGRATION_TEST: ""
+  TESTS_IN_PARALLEL: "8"
+  XDIST_PARALLEL_ARG: "8" #TODO: Unify these two parameters
+  PUBLISH_ARTIFACTS: ""
+  STOP_SLAVE: ""
+  CLEAN_BUILD_CACHE: ""
 
 stages:
+  - init
   - build
+  - test
+
+init_workspace:
+  stage: init
+  image: alpine:latest
+  tags:
+    - mender-qa-slave
+  script:
+    - export WORKSPACE=${CI_PROJECT_DIR}/..
+    - apk --update --no-cache add git openssh
+    # Prepare SSH keys
+    - eval $(ssh-agent -s)
+    - echo "$SSH_PRIVATE_KEY" | tr -d '\r' | ssh-add - > /dev/null
+    - mkdir -p ~/.ssh
+    - chmod 700 ~/.ssh
+    - ssh-keyscan github.com >> ~/.ssh/known_hosts
+    # Clone all repos in WORKSPACE
+    - cd ${WORKSPACE}
+    - git init . && git remote add origin https://github.com/mendersoftware/poky || true
+    - git fetch && git checkout -f origin/${POKY_REV}
+    - echo POKY_REV_GIT_SHA=$(git rev-parse HEAD) > ${CI_PROJECT_DIR}/build_revisions.env
+    - git clone https://github.com/mendersoftware/meta-mender || true
```

Comment:
Why all the `|| true`? Shouldn't it error if the clone fails?

Reply:
The folders might already exist when the slaves are reused, so the clone will fail. If there is an error with the repo itself, we will still catch it in the next line, right?

Please review well also the next line, where I do the fetch pull/xxx/head into local branch pr. Will git call be robust when reusing slaves? I haven't verified it.

Reply:
Oh, I thought that since it's a container, it would always be fresh. But ok, I understand the approach then.

Reply:
I thought so too... But experiments have showed otherwise.

What about this kind of checkouts: `(cd meta-mender && git fetch -u -f origin ${META_MENDER_REV}:pr && git checkout pr)`. Do they look good to you?

Reply:
wouldn't a `rm -rf meta_mender && git clone -b ${META_MENDER_REV} --single-branch https://github.com/mendersoftware/meta-mender.git` also work? then we would reduce the 3 lines per project + speed up the cloning, by only focusing on the single-branch

- git clone https://github.com/mendersoftware/mender go/src/github.com/mendersoftware/mender || true
- (cd go/src/github.com/mendersoftware/mender && git fetch -u -f origin ${MENDER_REV}:pr && git checkout pr)

Reply:
Don't use single-branch. There is a lot of Git magic being used in various places that rely on a full repository. One example is the extremely long running Uboot compile test (~2h for a single test), which uses branches to find out whether the feature has been recently updated or not, and skipping based on this.